cdap-app-fabric/src/main/java/io/cdap/cdap/app/runtime/AbstractProgramRuntimeService.java

```
@@ -169,17 +181,37 @@ public final RuntimeInfo run(ProgramDescriptor programDescriptor, ProgramOptions
-      ? programRunnerFactory
-      : Optional.ofNullable(remoteProgramRunnerFactory).orElseThrow(UnsupportedOperationException::new)
-    ).create(programId.getType());
+    ProgramRunnerFactory progRunnerFactory = (clusterMode == ClusterMode.ON_PREMISE) ? programRunnerFactory :
+      Optional.ofNullable(remoteProgramRunnerFactory).orElseThrow(UnsupportedOperationException::new);
+    ProgramRunner runner = progRunnerFactory.create(programId.getType());
 
     File tempDir = createTempDirectory(programId, runId);
     AtomicReference<Runnable> cleanUpTaskRef = new AtomicReference<>(createCleanupTask(tempDir, runner));
     DelayedProgramController controller = new DelayedProgramController(programRunId);
     RuntimeInfo runtimeInfo = createRuntimeInfo(controller, programId, () -> cleanUpTaskRef.get().run());
     updateRuntimeInfo(runtimeInfo);
 
+    String peer = options.getArguments().getOption(ProgramOptionConstants.PEER_NAME);
+    if (peer != null) {
+      try {
+        // For tethered pipeline runs, fetch artifacts from ArtifactCacheService
+        String basePath = String.format("%s/peers/%s", Constants.Gateway.INTERNAL_API_VERSION_3, peer);
+        RemoteClient remoteClient = remoteClientFactory.createRemoteClient(
+          Constants.Service.ARTIFACT_CACHE_SERVICE,
+          RemoteClientFactory.NO_VERIFY_HTTP_REQUEST_CONFIG,
+          basePath);
+        RemoteArtifactRepositoryReader artifactRepositoryReader = new RemoteArtifactRepositoryReader(
+          locationFactory, remoteClient);
+        noAuthArtifactRepository = new RemoteArtifactRepository(cConf, artifactRepositoryReader,
+                                                                progRunnerFactory);
+      } catch (Exception e) {
+        controller.failed(e);
+        programStateWriter.error(programRunId, e);
+        LOG.error("Exception while trying to create remote client factory", e);
+        return runtimeInfo;
```

what happens after this is returned? Would it make sense to let the exception propagate up rather than return the runtimeInfo?

---

Existing code catches exceptions in this method, sets program controller in failed state and writes an error to program state writer - https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/app/runtime/AbstractProgramRuntimeService.java#L223. I have followed the same pattern here.

The caller also doesn't expect an exception to be thrown - https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java#L693